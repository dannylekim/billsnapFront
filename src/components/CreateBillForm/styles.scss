div.modal__create{
    overflow-y: auto;
    height: 50em;
}