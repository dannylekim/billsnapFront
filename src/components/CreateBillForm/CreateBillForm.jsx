import React from "react";
import "./styles.scss";

import {Button, Col, Container, Form, FormGroup, FormInput, Row, FormRadio, InputGroup, InputGroupAddon, InputGroupText} from "shards-react";

export const CreateBillForm = ({
  handleSubmitClick,
  handleAddClick,
  handleRemoveClick,
  onFormChange,
  onCategoryChange,
  hasErrors,
  errorMessage,
  addBillForm,
  balance = 0,
  tipFormat
}) => {
  return (
    <div>
      <Form>
        <FormGroup onChange={onFormChange}>
          <label htmlFor="#name"> BILL NAME </label>
          <FormInput name="name" id="#name" placeholder="ex: Bill #1"/>
          <Container>
            <Row>
              <Col>
                <label htmlFor="#paidBy"> PAID BY </label>
                <FormInput name="paidBy" id="#paidBy" placeholder="ex: Danny"/>
              </Col>
              <Col>
                <label htmlFor="#category"> CATEGORY </label>
                <FormInput name="category" id="#category" placeholder="ex: Restaurant"/>
              </Col>
            </Row>
          </Container>
          <label htmlFor="#company"> COMPANY </label>
          <FormInput name="company" id="#company" placeholder="ex: OnlyFans"/>
        </FormGroup>

        <FormGroup onChange={(event) => onCategoryChange(event, "item")}>
          <Container>
            <Row>
              <Col>Items</Col>
              <Col>Price</Col>
              <Col />
            </Row>
            {addBillForm.items.map((item, key) => (
              <div>
                <Row key={key}>
                  <Col> {item.name}</Col>
                  <Col> {item.cost}</Col>
                  <Col> 
                    <div onClick={() => handleRemoveClick("item", key)}> - </div>
                  </Col>
                </Row>
                <hr />
              </div>
            ))}
            <Form id="itemInputs">
              <Row>
                <Col>
                  <FormInput name="name" placeholder="ex: Goat cheese"/>
                </Col>
                <Col>
                  <InputGroup>
                    <InputGroupAddon type="prepend">
                      <InputGroupText>$</InputGroupText>
                    </InputGroupAddon>
                    <FormInput name="cost" type="number" min="0" placeholder="ex: 37.43" />
                  </InputGroup>
                </Col>
                <Button onClick={() => handleAddClick("item")}>Add</Button>
              </Row>
            </Form>
          </Container>
        </FormGroup>
        
        <FormGroup onChange={(event) => onCategoryChange(event, "tax")}>
          <Container>
            <Row>
              <Col>Tax Name</Col>
              <Col>Percentage</Col>
              <Col />
            </Row>
            {addBillForm.taxes.map((tax, key) => (
              <div>
                <Row key={key}>
                  <Col> {tax.name}</Col>
                  <Col> {tax.percentage}</Col>
                  <Col> 
                    <div onClick={() => handleRemoveClick("tax", key)}> - </div>
                  </Col>
                </Row>
                <hr />
              </div>
            ))}
            <Form id="taxInputs">
              <Row>
                <Col>
                  <FormInput name="name" placeholder="ex: TPS"/>
                </Col>
                <Col>
                  <InputGroup>
                    <InputGroupAddon type="prepend">
                      <InputGroupText>%</InputGroupText>
                    </InputGroupAddon>
                    <FormInput name="percentage" type="number" min="0" placeholder="ex: 5" />
                  </InputGroup>
                </Col>
                <Button onClick={() => handleAddClick("tax")}>Add</Button>
              </Row>
            </Form>
          </Container>
        </FormGroup>

        <FormGroup onChange={onFormChange}>
          <label htmlFor="#tip"> TIP </label>
          <FormRadio 
            inline 
            checked={tipFormat} 
            onChange={(event) => onCategoryChange(event, "tipFormat")} >Amount</FormRadio>
          <FormRadio 
            inline 
            checked={!tipFormat}
            onChange={(event) => onCategoryChange(event, "tipFormat")} >Percentage</FormRadio>
          
          <Form id="tipInputs">
            {tipFormat ? (
              <InputGroup>
                <InputGroupAddon type="prepend">
                  <InputGroupText> $</InputGroupText>
                </InputGroupAddon>
                <FormInput name="tipAmount" type="number" min="0" placeholder="ex: 3.69" />
              </InputGroup>
            ) : (
              <InputGroup>
                <InputGroupAddon type="prepend">
                  <InputGroupText>%</InputGroupText>
                </InputGroupAddon>
                <FormInput name="tipPercent" type="number" min="0" placeholder="ex: 25" />
              </InputGroup>
            )}
          </Form>
        </FormGroup>

        <div>Total amount to split: {balance}</div>
        
        <FormGroup onChange={(event) => onCategoryChange(event, "account")}>
          <Container>
            <Row>
              <Col>Invite Users</Col>
              <Col />
            </Row>
            {addBillForm.accountsList.map((email, key) => (
              <div>
                <Row key={key}>
                  <Col> {email}</Col>
                  <Col> 
                    <div onClick={() => handleRemoveClick("account", key)}> - </div>
                  </Col>
                </Row>
                <hr />
              </div>
            ))}
            <Form id="accountInputs">
              <Row>
                <Col>
                  <FormInput name="email" placeholder="friend@email.com"/>
                </Col>
                <Button onClick={() => handleAddClick("account")}>Add</Button>
              </Row>
            </Form>
          </Container>
        </FormGroup>
      </Form>
      <Button size="md" pill onClick={handleSubmitClick} name="confirm">
        Confirm a
      </Button>
    </div>
  );
};
