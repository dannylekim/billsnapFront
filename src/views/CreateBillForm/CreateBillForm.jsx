import React, {Component} from "react";
import "./styles.scss";

class CreateBillFormContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }
}

export const CreateBillForm = (props) => {
  return <div/>;
};

export default CreateBillFormContainer;
