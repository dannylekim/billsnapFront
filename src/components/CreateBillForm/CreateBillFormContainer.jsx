import React, {Component} from "react";
import {CreateBillForm} from "./CreateBillForm";
import {Button, Modal, ModalBody, ModalHeader} from "shards-react";

class CreateBillFormContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      addBillForm: {
        name: "",
        category: "",
        company: "",
        items: [],
        accountsList: [],
        tipAmount: 0,
        tipPercent: 0,
        taxes: []
      },
      balance: 0,
      isOpen: false,
      hasErrors: {},
      errorMessage: "",
      isLoading: false,
      itemBuffer: {name: "", cost: 0},
      taxBuffer: {name: "", percentage: 0},
      accountBuffer: "",
      tipFormat: true
    };
    this.handleSubmitClick = this.handleSubmitClick.bind(this);
    this.handleAddClick = this.handleAddClick.bind(this);
    this.handleRemoveClick = this.handleRemoveClick.bind(this);
    this.onFormChange = this.onFormChange.bind(this);
    this.onCategoryChange = this.onCategoryChange.bind(this);
    this.handleErrorResponse = this.handleErrorResponse.bind(this);
    this.toggleModal = this.toggleModal.bind(this);
    this.handleCorrectTipFormat = this.handleCorrectTipFormat.bind(this);
  }

  /**
   * @function onFormChange for non calculating item
   * @description event handling function, handles the changes in the input fields.
   * @param {Event} event
   */
  onFormChange = (event) => {
    const { name, value } = event.target;
    this.setState((prev) => ({
      addBillForm: {
        ...prev.addBillForm,
        [name]: value,
      },
    }));
  }

  /**
   *
   *
   * @param event
   * @param isAdd
   */
  onCategoryChange = (event, category) => {
    const { name, value } = event.target;
    switch (category) {
      case "item":
        this.setState((prev) => ({
          itemBuffer: {
            ...prev.itemBuffer,
            [name]: value
          }
        }));
        break;

      case "tax":
        this.setState((prev) => ({
          taxBuffer: {
            ...prev.taxBuffer,
            [name]: value
          }
        }));
        break;

      case "tipFormat":
        this.setState({
          tipFormat: !this.state.tipFormat,
        });
        break;

      case "account":
        this.setState(() => ({
          accountBuffer: value
        }));
        break;
    }
  };

  /**
   * @function handleCorrectTipFormat
   */
  handleCorrectTipFormat = () => {
    if (this.state.tipFormat) {
      this.setState((prev) => ({
        addBillForm: {
          ...prev.addBillForm,
          tipPercent: undefined,
        },
      }));
    } else  {
      this.setState((prev) => ({
        addBillForm: {
          ...prev.addBillForm,
          tipAmount: undefined,
        },
      }));
    }
  };

  /**
   * @function handleSubmitClick
   * @description Handles the submission by calling the create bill endpoint.
   * handle the response with handleResponse function.
   * @param {Event} event
   */
  handleSubmitClick = async (event) => {
    event.preventDefault();
    try {
      this.setState({
        isLoading: true,
      });

      this.handleCorrectTipFormat();
      
      console.log(this.state.addBillForm);
      const newBill = await this.props.createNewBill(this.state.addBillForm);
      this.setState({
        isOpen: true,
      });
      return newBill;
    } catch (error) {
      this.handleErrorResponse(error);
    } finally {
      this.setState({
        isLoading: false,
      });
    }
  };

  /**
   * @function handleItemAddClick
   */
  handleAddClick = (category) => {
    switch (category) {
      case "item":
        const itemBuffer = this.state.itemBuffer;
        if (itemBuffer.name && itemBuffer.cost) {
          this.setState({
            items: this.state.addBillForm.items.push({ ...itemBuffer }),
            balance: this.state.balance + itemBuffer.cost,
            itemBuffer: { name: "", cost: 0 }
          });
        }
        break;

      case "tax":
        const taxBuffer = this.state.taxBuffer;
        if (taxBuffer.name && taxBuffer.percentage) {
          this.setState({
            taxes: this.state.addBillForm.taxes.push({ ...taxBuffer }),
            taxBuffer: {name: "", percentage: 0}
          });
        }
        break;

      case "account":
        let accBuffer = this.state.accountBuffer;
        if (accBuffer) {
          this.setState({
            accountsList: this.state.addBillForm.accountsList.push(accBuffer),
            accountBuffer: ""
          })
        }
        break;
    }
  }

    /**
   * @function handleItemRemoveClick
   */
  handleRemoveClick = (category, index) => {
    switch (category) {
      case "item":
        const items = this.state.addBillForm.items;
        delete items[index];
        this.setState({ items: items });
        break;

      case "tax":
        const taxes = this.state.addBillForm.taxes;
        delete taxes[index];
        this.setState({ taxes: taxes });
        break;

      case "account":
        const accounts = this.state.addBillForm.accountsList;
        delete accounts[index];
        this.setState({ accountsList: accounts });
        break;
    }
  }

  /**
   * Toggles the modal from open to close
   */
  toggleModal = () => {
    this.setState((prev) => ({
      isOpen: !prev.isOpen,
    }));
  };

  handleErrorResponse = (response) => {
    console.log("wew")
    console.log(response.errors)
  };

  render() {
    return (
      <div>
        <Button onClick={this.toggleModal}> + Add Bill</Button>
        <Modal size="lg" open={this.state.isOpen} toggle={this.toggleModal}>
          <ModalHeader>Bill details</ModalHeader>
          <ModalBody>
            <CreateBillForm
              handleSubmitClick={this.handleSubmitClick}
              handleAddClick={this.handleAddClick}
              handleRemoveClick={this.handleRemoveClick}
              onFormChange={this.onFormChange}
              onCategoryChange={this.onCategoryChange}
              hasErrors={this.state.hasErrors}
              errorMessage={this.errorMessage}
              addBillForm={this.state.addBillForm}
              balance={this.state.balance}
              tipFormat={this.state.tipFormat}
              itemBuffer={this.state.itemBuffer}
              taxBuffer={this.state.taxBuffer}
              accountBuffer={this.state.accountBuffer}
            />
          </ModalBody>
        </Modal>
      </div>
    );
  }
}

export default CreateBillFormContainer;
