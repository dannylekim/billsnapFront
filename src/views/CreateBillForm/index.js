import CreateBillForm from "./CreateBillForm";

export default CreateBillForm;