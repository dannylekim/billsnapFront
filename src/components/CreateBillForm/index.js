import CreateBillFormContainer from "./CreateBillFormContainer";
import {connect} from "react-redux";
import {createNewBill} from "../../redux/actions/billActions";

const mapDispatchToProps = (dispatch) => ({
  createNewBill: (createBillParam) => dispatch(createNewBill(createBillParam)),
});

export default connect(null, mapDispatchToProps)(CreateBillFormContainer);
