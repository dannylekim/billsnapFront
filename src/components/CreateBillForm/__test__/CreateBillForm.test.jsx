import React from "react";
import ReactDOM from "react-dom";
import {CreateBillForm} from "../CreateBillForm";
import { DEFAULT_ERRORS, DEFAULT_ALERT_MESSAGE} from "../CreateBillFormContainer";

const addBillForm = {
    name: "",
    category: "",
    company: "",
    items: [{name: "cheese", cost: 10}],
    accountsList: ["email"],
    tipAmount: 0,
    tipPercent: 0,
    taxes: [{name: "tps", percentage: 11}]
}

describe("CreateBillForm", () => {

    describe("render", () => {
        it("renders without crashing", () => {
            const div = document.createElement("div");
            ReactDOM.render(<CreateBillForm
                handleSubmitClick={jest.fn}
                handleAddClick={jest.fn}
                handleRemoveClick={jest.fn}
                onFormChange={jest.fn}
                hasErrors={DEFAULT_ERRORS}
                errorMessage={""}
                alertMessage={DEFAULT_ALERT_MESSAGE}
                addBillForm={addBillForm}
                totalBalance={0}
                tipFormat={true}
                itemBuffer={{name: "", cost: 0}}
                taxBuffer={{name: "", percentage: 0}}
                accountBuffer={""}
                inputList={[]}/>, div);
        });
    });

    describe("snapshots ðŸ“¸", () => {
        it("CreateBillFormContainer should match snap shot", () => {
            matches(<CreateBillForm
                handleSubmitClick={jest.fn}
                handleAddClick={jest.fn}
                handleRemoveClick={jest.fn}
                onFormChange={jest.fn}
                hasErrors={DEFAULT_ERRORS}
                errorMessage={""}
                alertMessage={DEFAULT_ALERT_MESSAGE}
                addBillForm={addBillForm}
                totalBalance={0}
                tipFormat={true}
                itemBuffer={{name: "", cost: 0}}
                taxBuffer={{name: "", percentage: 0}}
                accountBuffer={""}
                inputList={[]}/>
            );
        });
    });

});
