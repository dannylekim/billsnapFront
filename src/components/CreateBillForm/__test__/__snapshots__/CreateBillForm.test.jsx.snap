// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateBillForm snapshots ðŸ“¸ CreateBillFormContainer should match snap shot 1`] = `
<div>
  <form
    className=""
  >
    <div
      className="form-group"
      onChange={[Function]}
    >
      <label
        htmlFor="#name"
      >
         BILL NAME 
      </label>
      <input
        className="form-control"
        id="name"
        name="name"
        placeholder="ex: Bill #1"
      />
      <div
        className="container"
      >
        <div
          className="row"
        >
          <div
            className="display__left col-lg-5"
          >
            <label
              htmlFor="#paidBy"
            >
               PAID BY 
            </label>
            <input
              className="form-control"
              id="paidBy"
              name="paidBy"
              placeholder="ex: Danny"
            />
          </div>
          <div
            className="col-lg-5"
          >
            <label
              htmlFor="#category"
            >
               CATEGORY 
            </label>
            <input
              className="form-control"
              id="category"
              name="category"
              placeholder="ex: Restaurant"
            />
          </div>
          <div
            className="col-lg-1"
          />
        </div>
      </div>
      <label
        htmlFor="#company"
      >
         COMPANY 
      </label>
      <input
        className="form-control"
        id="company"
        name="company"
        placeholder="ex: OnlyFans"
      />
    </div>
    <div
      className="form-group"
      id="items"
    >
      <div
        className="container"
      >
        <div
          className="row"
        >
          <div
            className="display__left col"
          >
            <label>
              <strong>
                Items
              </strong>
            </label>
          </div>
          <div
            className="col"
          >
            <label>
              PRICE
            </label>
          </div>
          <div
            className="col"
          />
        </div>
        <div>
          <div
            className="element__display row"
          >
            <div
              className="display__left col"
            >
               
              cheese
            </div>
            <div
              className="col"
            >
               
              10
            </div>
            <div
              className="col"
            >
              <div
                className="remove__button"
                onClick={[Function]}
              >
                 - 
              </div>
            </div>
          </div>
          <hr />
        </div>
        <div
          className="row"
        >
          <div
            className="display__left col"
          >
            <input
              className="form-control"
              name="name"
              onChange={[Function]}
              placeholder="ex: Goat cheese"
              value=""
            />
          </div>
          <div
            className="col"
          >
            <div
              className="input-group"
            >
              <div
                className="input-group-prepend"
              >
                <span
                  className="input-group-text"
                >
                  $
                </span>
              </div>
              <input
                className="form-control"
                min="0"
                name="cost"
                onChange={[Function]}
                placeholder="ex: 37.43"
                type="number"
                value={0}
              />
            </div>
          </div>
          <div
            className="col"
          >
            <button
              className="btn btn-primary"
              onClick={[Function]}
              type="button"
            >
              Add
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="form-group"
    >
      <div
        className="container"
      >
        <div
          className="row"
        >
          <div
            className="col-lg-4"
          >
            <label
              htmlFor="#tip"
            >
               
              <strong>
                TIP
              </strong>
            </label>
          </div>
          <div
            className="col-lg-5"
          >
            <label
              className="custom-control custom-radio custom-control-inline"
            >
              <input
                checked={true}
                className="custom-control-input"
                id="dr-radio-73llaHQYDC"
                onChange={[Function]}
                type="radio"
              />
              <label
                aria-hidden="true"
                className="custom-control-label"
                id="dr-radio-73llaHQYDC"
                onClick={[Function]}
              />
              <span
                className="custom-control-description"
              >
                Amount
              </span>
            </label>
            <label
              className="custom-control custom-radio custom-control-inline"
            >
              <input
                checked={false}
                className="custom-control-input"
                id="dr-radio-EPamlvD3lj"
                onChange={[Function]}
                type="radio"
              />
              <label
                aria-hidden="true"
                className="custom-control-label"
                id="dr-radio-EPamlvD3lj"
                onClick={[Function]}
              />
              <span
                className="custom-control-description"
              >
                Percentage
              </span>
            </label>
          </div>
        </div>
      </div>
      <div
        className="input-group"
      >
        <div
          className="input-group-prepend"
        >
          <span
            className="input-group-text"
          >
             $
          </span>
        </div>
        <input
          className="form-control"
          id="tip"
          min="0"
          name="tipAmount"
          onChange={[Function]}
          placeholder="ex: 3.69"
          type="number"
          value={0}
        />
      </div>
    </div>
    <div
      className="form-group"
      id="taxes"
    >
      <div
        className="container"
      >
        <div
          className="row"
        >
          <div
            className="display__left col"
          >
            <label>
              <strong>
                Tax Name
              </strong>
            </label>
          </div>
          <div
            className="col"
          >
            <label>
              PERCENTAGE
            </label>
          </div>
          <div
            className="col"
          />
        </div>
        <div>
          <div
            className="element__display row"
          >
            <div
              className="display__left col"
            >
               
              tps
            </div>
            <div
              className="col"
            >
               
              11
            </div>
            <div
              className="col"
            >
              <div
                className="remove__button"
                onClick={[Function]}
              >
                 - 
              </div>
            </div>
          </div>
          <hr />
        </div>
        <div
          className="row"
        >
          <div
            className="display__left col"
          >
            <input
              className="form-control"
              name="name"
              onChange={[Function]}
              placeholder="ex: TPS"
              value=""
            />
          </div>
          <div
            className="col"
          >
            <div
              className="input-group"
            >
              <div
                className="input-group-prepend"
              >
                <span
                  className="input-group-text"
                >
                  %
                </span>
              </div>
              <input
                className="form-control"
                max="100"
                min="0"
                name="percentage"
                onChange={[Function]}
                placeholder="ex: 5"
                type="number"
                value={0}
              />
            </div>
          </div>
          <div
            className="col"
          >
            <button
              className="btn btn-primary"
              onClick={[Function]}
              type="button"
            >
              Add
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="balance container"
    >
      <div
        className="row"
      >
        <div
          className="col-lg-4"
        >
          <strong>
            Total amount to split:
          </strong>
        </div>
        <div
          className="col-lg-8"
        >
          <span
            id="balance__amount"
          >
            0
             $
          </span>
        </div>
      </div>
    </div>
    <div
      className="form-group"
    >
      <div
        className="container"
      >
        <div
          className="row"
        >
          <div
            className="col-lg-4"
          >
            <strong>
              Invite Users
            </strong>
          </div>
          <div
            className="col-lg-8"
          />
        </div>
        <div>
          <div
            className="element__display row"
          >
            <div
              className="display__left col"
            >
               
              email
            </div>
            <div
              className="col"
            >
              <div
                className="remove__button"
                onClick={[Function]}
              >
                 - 
              </div>
            </div>
          </div>
          <hr />
        </div>
        <div
          className="row"
        >
          <div
            className="display__left col-lg-8"
          >
            <input
              className="form-control"
              id="account"
              name="email"
              onChange={[Function]}
              placeholder="friend@email.com"
              value=""
            />
          </div>
          <div
            className="col-lg-4"
          >
            <button
              className="btn btn-primary"
              onClick={[Function]}
              type="button"
            >
              Add
            </button>
          </div>
        </div>
      </div>
    </div>
  </form>
  <button
    className="btn btn-primary btn-md btn-pill"
    name="confirm"
    onClick={[Function]}
    type="button"
  >
    Confirm
  </button>
</div>
`;
